tools/clang/test/CodeGenHLSL/batch/declarations/resources/byteaddressbuffers/load_type_shapes_sm60.hlsl

```diff
@@ -0,0 +1,38 @@
+// RUN: %dxc -E main -T vs_6_0 %s | FileCheck %s
+
+// Tests that ByteAddressBuffer.Load<T> works with all type shapes with SM 6.0
+
+struct S { int i; float f; };
```

Comment:
Hi @MrTrillian 
One more question...
```hlsl
struct S {
  uint16_t i;
  float f;
};

buf.Load<S[2]>(/*address*/ 0);
```

How do you handle the load at bit-offset 16? Can you load at bit-offset 16? or do you load two 32-bit values, and then perform operations to separate out the [0-15] bits and [16-47] bits?

Reply:
@MrTrillian I'm not sure if you saw my question here.
I'm wondering how you handle 16-bit and 64-bit loads at non-zero bit-offsets on DXIL side?

Reply:
Sorry Ehsan! We have a `rawBufferLoad.i16` intrinsic in DXIL which loads up to 4 subsequent 16-bit values and it has an alignment parameter. The rest is up to the backend. Our spec says that "The call respects SM5.1 OOB and alignment rules." but I'll have to defer to @tex3d for the exact meaning of that.

I've checked with some of our tests and we emit an alignment value of `2`, so I presume that in the case of loading a `uint16_t2` at a dynamic offset, a dword-centric backend would have to generate a conditional second load in case the second uint16_t falls in a different dword, and do the bit shifts as needed.

64-bit types work similarly, we have a `rawBufferLoad.i64` intrinsic. Your question made me think and we might have a bug here though: #2263

Reply:
Thanks for the explanation @MrTrillian.

So I think my question boils down to padding. I'm not sure whether or not loading and storing is done with padding to conform to some alignment rules?

For example, for the following structure, after loading `a`, do the bits for `b` start right after `a` (i.e. everything is tightly packed), or do they start at the next aligned offset (i.e. padding is added)?

For
```hlsl
ByteAddressBuffer b1;
RWByteAddressBuffer b2;

struct S {
  float16_t a;
  double b;
};

[numthreads(64, 1, 1)]
void main(uint3 tid : SV_DispatchThreadId)
{
  S myS = b1.Load<S>(tid.x);
  b2.Store<S>(0, myS);
}
```
I get:
```
define void @main() {
  %b2_UAV_rawbuf = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 1, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %b1_texture_rawbuf = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 0, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %1 = call i32 @dx.op.threadId.i32(i32 93, i32 0)  ; ThreadId(component)
  %RawBufferLoad1 = call %dx.types.ResRet.f16 @dx.op.rawBufferLoad.f16(i32 139, %dx.types.Handle %b1_texture_rawbuf, i32 %1, i32 undef, i8 1, i32 2)  ; RawBufferLoad(srv,index,elementOffset,mask,alignment)
  %2 = extractvalue %dx.types.ResRet.f16 %RawBufferLoad1, 0
  %3 = add i32 %1, 8
  %4 = call %dx.types.ResRet.i32 @dx.op.rawBufferLoad.i32(i32 139, %dx.types.Handle %b1_texture_rawbuf, i32 %3, i32 undef, i8 3, i32 8)  ; RawBufferLoad(srv,index,elementOffset,mask,alignment)
  %5 = extractvalue %dx.types.ResRet.i32 %4, 0
  %6 = extractvalue %dx.types.ResRet.i32 %4, 1
  %7 = call double @dx.op.makeDouble.f64(i32 101, i32 %5, i32 %6)  ; MakeDouble(lo,hi)
  call void @dx.op.rawBufferStore.f16(i32 140, %dx.types.Handle %b2_UAV_rawbuf, i32 0, i32 undef, half %2, half undef, half undef, half undef, i8 1, i32 2)  ; RawBufferStore(uav,index,elementOffset,value0,value1,value2,value3,mask,alignment)
  %8 = call %dx.types.splitdouble @dx.op.splitDouble.f64(i32 102, double %7)  ; SplitDouble(value)
  %9 = extractvalue %dx.types.splitdouble %8, 0
  %10 = extractvalue %dx.types.splitdouble %8, 1
  call void @dx.op.rawBufferStore.i32(i32 140, %dx.types.Handle %b2_UAV_rawbuf, i32 8, i32 undef, i32 %9, i32 %10, i32 undef, i32 undef, i8 3, i32 8)  ; RawBufferStore(uav,index,elementOffset,value0,value1,value2,value3,mask,alignment)
  ret void
}
```

* Based on my limited understanding of DXIL, I think ` %3 = add i32 %1, 8` (=`tid.x + 8`) means loading `b` starts 8 bytes after the address of `a`. Which means `a` is padded with 48 bits that are ignored. Am I right?

* Can you point to a doc/link what describes which alignment rules we should follow?

Thanks again!

Reply:
Yes, there is padding. Structures and arrays in HLSL have two layouts: the "natural" layout and the constant buffer layout.

The constant buffer layout is a legacy from DXBC and is only relevant for accesses through `ConstantBuffer<Struct>` and follows [these rules](https://docs.microsoft.com/en-us/windows/desktop/direct3dhlsl/dx-graphics-hlsl-packing-rules).

The "natural" layout more closely ressembles C++, though I can't find official documentation for it :/. Basically, the rules are:
- Bools are like ints in memory (4 bytes)
- Scalars are aligned in accordance with their size (int16_t => 2, int => 4, int64_t => 8)
- Vectors/matrices are aligned like their element type
- Arrays are aligned like their element type
- Structs are aligned like their field with the biggest alignment
- Empty structs have size zero

This matches the field offsets reported by LLVM once lowered to the IR. I'm not sure if it matches the clang field offsets in the presence of vector/matrices of bools.

Whenever structs are loaded/stored from/to non-constant buffers, be it `StructuredBuffer<S>.operator[]` or `ByteAddressBuffer.Load<S>`, the "natural" layout is used.

Reply:
Got it! Thanks for your quick response. I think this covers all the information I needed 👍